File: src/runtime/typescript/src/jobs-cancel-route.ts
/**
 * Mount the cancel HTTP route on the agent's HTTP server (Phase 1 —
 * MeshJob substrate).
 *
 * Mirrors Python's
 * `_mcp_mesh.pipeline.mcp_startup.jobs_cancel_route.JobsCancelRouteStep`.
 *
 * Per `MESHJOB_DESIGN.org` "Wire Protocol / New endpoints" /
 * "Cancellation": when the registry receives a cancel request for a
 * job whose owner is alive, it forwards the call to the owner
 * replica's `POST /jobs/{job_id}/cancel` HTTP route. That route fires
 * the in-process cancel token registered by the inbound tool wrapper
 * (or the claim worker) so any outbound HTTP calls under the active
 * `with_job_async` scope abort.
 *
 * The route is registered on FastMCP's underlying Hono app. FastMCP
 * exposes the Hono instance via `server.getApp()` after the server has
 * started; calling that BEFORE start would throw. We register at the
 * front of the routes table so the explicit `/jobs/:job_id/cancel`
 * matches before any catch-all FastMCP routes (Hono matches routes in
 * registration order by default, so calling `app.post(...)` after the
 * server starts is sufficient — Hono inserts at the end, but the
 * FastMCP catch-alls only match the configured `endpoint` path which
 * defaults to `/mcp` — there's no overlap).
 */
import type { FastMCP } from "fastmcp";
import { cancelActiveJob } from "@mcpmesh/core";

/**
 * Register `POST /jobs/:job_id/cancel` on the FastMCP server's Hono
 * app. Returns `true` iff the route was registered (i.e. the FastMCP
 * server exposes a Hono instance). Logs a warning and returns `false`
 * on any failure — cancel is a best-effort signal; missing the route
 * just means the registry's forward attempt 404s and the cancel
 * falls back to lease-expiry.
 */
export function registerCancelRoute(server: FastMCP): boolean {
  let app: ReturnType<FastMCP["getApp"]> | null = null;
  try {
    app = server.getApp();
  } catch (err) {
    console.warn(
      "[mesh-jobs] FastMCP.getApp() unavailable — cancel route not registered:",
      err,
    );
    return false;
  }
  if (!app) {
    console.warn(
      "[mesh-jobs] FastMCP.getApp() returned null — cancel route not registered",
    );
    return false;
  }

  try {
    app.post("/jobs/:job_id/cancel", async (c) => {
      const jobId = c.req.param("job_id");
      let cancelled = false;
      try {
        cancelled = Boolean(cancelActiveJob(jobId));
      } catch (err) {
        // The napi binding shouldn't throw under normal conditions;
        // log + report cancelled=false so the registry's forwarder
        // gets a deterministic answer.
        console.warn(
          `[mesh-jobs] cancelActiveJob napi raised for job=${jobId}:`,
          err,
        );
      }
      console.log(
        `[mesh-jobs] cancel route: job_id=${jobId} cancelled=${cancelled}`,
      );
      return c.json({ cancelled, job_id: jobId });
    });
    return true;
  } catch (err) {
    console.warn("[mesh-jobs] failed to register cancel route:", err);
    return false;
  }
}

File: src/runtime/python/_mcp_mesh/pipeline/mcp_startup/jobs_cancel_route.py
"""Register the inbound cancel HTTP route on the agent's FastAPI app
(Phase 1 — MeshJob substrate).

Per ``MESHJOB_DESIGN.org`` "Wire Protocol / New endpoints" /
"Cancellation": when the registry receives a cancel request for a job
whose owner is alive, it forwards the call to the owner replica's
``POST /jobs/{job_id}/cancel`` HTTP route. That route fires the
in-process cancel token registered by the inbound tool wrapper (or the
claim worker) so any outbound HTTP calls under the active
:func:`mcp_mesh_core.with_job_async` scope abort.

The route is registered on the FastAPI app prepared by
:class:`FastAPIServerSetupStep`, AFTER that step has stored the app in
the pipeline context. Returning ``{"cancelled": True/False}`` tells the
registry whether a token was found in this process — useful for
diagnostics when a forwarded cancel races with the owning job
finishing.

Route ordering note (#bug 4)
----------------------------

:class:`FastAPIServerSetupStep` mounts the FastMCP HTTP app at the
root path (``app.mount("", fastmcp_app)``). Starlette resolves routes
in registration order and ``Mount("")`` matches *every* path — so an
explicit route added AFTER the mount via ``app.add_api_route`` is
shadowed and yields 404. To work around that, this step inserts the
cancel route at the FRONT of ``app.router.routes`` so it's matched
before the catch-all FastMCP mount.
"""

from __future__ import annotations

import logging
from typing import Any

from ..shared import PipelineResult, PipelineStatus, PipelineStep

logger = logging.getLogger(__name__)


class JobsCancelRouteStep(PipelineStep):
    """Add ``POST /jobs/{job_id}/cancel`` to the agent's FastAPI app.

    Runs AFTER :class:`FastAPIServerSetupStep` (which prepares /
    discovers the FastAPI app) and BEFORE the orchestrator hands the
    app to uvicorn (so the route is part of the served app).
    """

    def __init__(self) -> None:
        super().__init__(
            name="jobs-cancel-route",
            required=False,  # Best-effort: failing must not block startup.
            description=(
                "Register POST /jobs/{job_id}/cancel on the agent's "
                "FastAPI app (forwards to mcp_mesh_core.cancel_active_job)"
            ),
        )

    async def execute(self, context: dict[str, Any]) -> PipelineResult:
        result = PipelineResult(message="MeshJob cancel route registration")

        app = context.get("fastapi_app")
        if app is None:
            result.status = PipelineStatus.SKIPPED
            result.message = (
                "MeshJob cancel route skipped: no FastAPI app in context "
                "(HTTP transport disabled?)"
            )
            self.logger.info("⚠️ %s", result.message)
            return result

        # Late import: keep the SDK importable in test environments
        # where the native extension isn't built. The route still gets
        # registered (so HTTP 404 doesn't surface to the registry); it
        # just always returns ``cancelled=False``.
        try:
            from mcp_mesh_core import cancel_active_job
        except Exception as e:
            self.logger.warning(
                "MeshJob cancel route: mcp_mesh_core.cancel_active_job "
                "unavailable (%s); registering a stub route that always "
                "reports cancelled=false",
                e,
            )
            cancel_active_job = lambda _job_id: False  # noqa: E731

        async def cancel_job_route(job_id: str) -> dict[str, Any]:
            """Fire the local cancel token for ``job_id`` if present."""
            cancelled = bool(cancel_active_job(job_id))
            self.logger.info(
                "📨 cancel route: job_id=%s cancelled=%s", job_id, cancelled
            )
            return {"cancelled": cancelled, "job_id": job_id}

        try:
            # Build the FastAPI route, then INSERT it at the front of the
            # router's routes list so the explicit path is matched before
            # FastAPIServerSetupStep's ``app.mount("", fastmcp_app)``
            # catch-all (#bug 4).
            from fastapi.routing import APIRoute

            route = APIRoute(
                path="/jobs/{job_id}/cancel",
                endpoint=cancel_job_route,
                methods=["POST"],
                tags=["mesh-jobs"],
                summary="Fire the in-process cancel token for the given job",
            )

            # Defensive: skip duplicate registration on hot-reload.
            existing_paths = {
                getattr(r, "path", None) for r in app.router.routes
            }
            if route.path not in existing_paths:
                app.router.routes.insert(0, route)
                self.logger.debug(
                    "Inserted MeshJob cancel route at routes[0] (front of "
                    "FastMCP catch-all mount)"
                )
            else:
                self.logger.debug(
                    "MeshJob cancel route already registered at %s; skipping",
                    route.path,
                )
        except Exception as e:
            result.status = PipelineStatus.FAILED
            result.message = f"MeshJob cancel route registration failed: {e}"
            result.add_error(str(e))
            self.logger.error("❌ %s", result.message)
            return result

        result.message = "Registered POST /jobs/{job_id}/cancel"
        self.logger.info("📨 %s", result.message)
        return result
